sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java

```diff
@@ -838,4 +884,50 @@ public final void prepareForProcessing() {}
    */
   @Override
   public void populateDisplayData(DisplayData.Builder builder) {}
+
+  /**
+   * A parameter that is accessible during {@link StartBundle @StartBundle}, {@link
+   * ProcessElement @ProcessElement} and {@link FinishBundle @FinishBundle} that allows the caller
+   * to register a callback that will be invoked after the bundle has been successfully completed
+   * and the runner has commit the output.
+   *
+   * <p>A common usage would be to perform any acknowledgements required by an external system such
+   * as acking messages from a message queue since this callback is only invoked after the output of
+   * the bundle has been durably persisted by the runner.
+   *
+   * <p>Note that a runner may make the output of the bundle available immediately to downstream
+   * consumers without waiting for finalization to succeed. For pipelines that are sensitive to
+   * duplicate messages, they must perform output deduplication in the pipeline.
+   */
+  // TODO: Add support for a deduplication PTransform.
+  @Experimental(Kind.SPLITTABLE_DO_FN)
+  public interface BundleFinalizer {
+    /**
+     * The provided function will be called after the runner successfully commits the output of a
+     * successful bundle. Throwing during finalization represents that bundle finalization may have
+     * failed and the runner may choose to attempt finalization again. The provided duration
+     * controls how long the finalization is valid for before it is garbage collected and no longer
+     * able to be invoked.
+     *
+     * <p>Note that finalization is best effort and it is expected that the external system will
+     * self recover state if finalization never happens or consistently fails. For example, a queue
+     * based system that requires message acknowledgement would replay messages if that
+     * acknowledgement was never received within the provided time bound.
+     *
+     * <p>See <a href="https://s.apache.org/beam-finalizing-bundles">Apache Beam Portability API:
+     * How to Finalize Bundles</a> for further details.
+     */
+    void afterBundleCommit(Duration finalizationDelayLimit, Callback callback);
```

Comment:
Similar to other the proto feedback: would this limit be more usable as an `Instant` rather than `Duration`? I assume in the general case the external system is providing a TTL, likely expressed as an absolute timestamp which should be passed through.

Reply:
Done in #7094